### PR TITLE
Faster parser for meta-language and smaller changes

### PR DESCRIPTION
**Pull request by sorig**
Here's the new `parse.py` version with both parsers. I've also addressed most of the issues discussed in the previous pull request #2722, including:

- Adding newlines at the end of example files
- cmake now automatically updates the typelist
- Rename gettypelist script
- Move README to wiki
- cmake generates a warning if PLY isn't installed
- Add unit tests to check that the two parsers are identical

I can't send pull requests to the wiki repository. Can you add the README.md to the wiki please? (I've already created a symlink to it)

### CODE REVIEW

File: cmake/FindPLY.cmake [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Finding PLY involves calling the Python interpreter
+execute_process(COMMAND "${PYTHON_EXECUTABLE}" "-c"
```
Comment (karlnapf) [inline cmake/FindPLY.cmake]: Did you write this or take it from somewhere?

File: examples/example-generation/CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,15 @@
 # Check which interfaces are turned on
 # Add CMakeLists file for each turned on interface
 
+# Generate type list
+execute_process(COMMAND sh ${CMAKE_CURRENT_SOURCE_DIR}/types/get_type_list.sh > ${CMAKE_CURRENT_SOURCE_DIR}/types/typelist)
```
Comment (karlnapf) [inline examples/example-generation/CMakeLists.txt]: nice!

File: examples/example-generation/tests/test_parse.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        program = """
+        CSVFile trainf("train.dat")
+        RealFeatures feats_train(trainf)
+        CSVFile testf("test.dat")
+        """
+
+        slowParsedProgram = self.slowParser.parse(program)
+        fastParsedProgram = self.fastParser.parse(program)
+
+        self.assertEqual(slowParsedProgram, fastParsedProgram)
```
Comment (karlnapf) [inline examples/example-generation/tests/test_parse.py]: great!

File: examples/example-generation/types/get_type_list.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+PATH_TO_SELF=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+MODULAR_DIR=$PATH_TO_SELF/../../../src/interfaces/modular/
+grep -Eor "(%rename\([^ ]*\))|(%template\([^ ]*\))" $MODULAR_DIR | grep -Eo "\([^ ]*\)" | grep -Eo "[^()]+" | grep -v "RealVector"
```
Comment (karlnapf) [inline examples/example-generation/types/get_type_list.sh]: scary ;)

File: cmake/FindPLY.cmake (hunk 1 above)
Comment (sorig) [inline cmake/FindPLY.cmake]: I copied from `cmake/FindJinja2.cmake` and adapted it to PLY.

**Comment by karlnapf** [comment]
Travis errors unrelated, merging

**Comment by karlnapf** [comment]
https://github.com/shogun-toolbox/shogun/wiki/Example_Generation

This works?

File: cmake/FindPLY.cmake (hunk 1 above)
Comment (karlnapf) [inline cmake/FindPLY.cmake]: ok cool!

**Comment by karlnapf** [comment]
Hey @sorig 
Could you update the readme file to explain the thing about PLY and also how to enable the examples from cmake. Also put a note in the cmake readme.
We will enable the examples in the builds soon
Thanks!

**Comment by karlnapf** [comment]
oh, and add this to the NEWS, too (might have to start a new section)

**Pull request merged**